Framework: People: Fix an issue where editing user with slug like usernames failed

ebinnion opened the pull request:
@enejb and I paired today and were able to come up with an alternative solution to #9408.

The original approach required adding some `people` specific login in to `getSiteFragment()`. And while it worked, what if we changed our routing in the future, or other sections had similar issues.

So, for our needs, we decided to flip the routing such that the username was before the site fragment, since `getSiteFragment()` checks the second-to-last position before checking the last segment. This handles the case for URLs like `/people/edit/example.com/user.name`. Then, to handle URLs like `/people/edit/123456/user.name`, we changed how `getSiteFragment()` works such that it will look in the second-to-last fragment for *both* a site fragment or an integer before moving on to the last segment.

To test:

- Checkout `update/people-edit-routing` branch
- Go to `/people/team/$site` where `$site` has a user with username that contains a `.`
- Click the user with the `.` in their username
- Ensure that you can properly view the user's profile
- Browse Calypso for breakage

enejb commented:
@mtias Do you have an opinion on this?

This PR looks good to me :)

aduth commented:
This seems to be a reasonable change, and both unit and user testing of existing edge cases appears to work well in my testing. 

LGTM 👍

mtias commented:
@ebinnion @enejb @aduth I'm not sure I understand the change, wasn't it the way it was supposed to work before? Didn't we supported slug or numeric IDs for sites as last or second to last params?

ebinnion commented:
Yes, we supported both slug and numeric IDs in both the last or second to last positions. But, we searched both positions for the slug before then searching both positions for a numeric ID.

The issues we ran into were:

- Since usernames can have a `.` within them, putting the username last in the route would cause `getSiteFragment` to return the username.
- So, we moved the slug to the second to last position. This works great, but what if we switched to using a numeric ID instead of a slug? It would fail.

This change, will check the second to last position for either a slug or numeric ID before checking the last position for a slug or numeric ID.

mtias commented:
Oh, didn't know we supported usernames with `.`. So `user.name` is a different user than `username`?

mtias commented:
> This change, will check the second to last position for either a slug or numeric ID before checking the last position for a slug or numeric ID.

I'm worried now that it could be possible we have a url like `/section/123/siteSlug` where the second param could be mistaken as a site?

ebinnion commented:
> Oh, didn't know we supported usernames with .. So user.name is a different user than username?

Correct. @enejb pointed it out to me and then pointed me to this function that sanitizes the user login: https://codex.wordpress.org/Function_Reference/sanitize_user

> I'm worried now that it could be possible we have a url like /section/123/siteSlug where the second param could be mistaken as a site?

That is also a possibility. That being said, there could also exist other issues where the second to last parameter may have a `.` (although, admittedly, probably edge case) and the old `getSiteFragment()` functionality fails there as well. The most fool-proof solution is probably to force the site fragment always being at the end of the route.

nb commented:
> The most fool-proof solution is probably to force the site fragment always being at the end of the route.

👍 This will take a lot of guessing out, which is definitely the best long-term option. Do we have a list somewhere with all routes that expect the site to be at second to last position?

mtias commented:
Anything that references a site specific resource doesn't make sense before the site fragment. An individual post is the most clear one: `/post/:site/:id` instead of `/post/:id/:site` because `:id` is not global. (We could try to look at changing it to be global, though, which would remove the need to have the site fragment in post urls.)

But you raise a good point for this particular case: since users can be part of more than one site, I'd expect `/people/edit/:user` to work and show me all the sites a user belongs to. That means the site fragment can very well come after since it's a filter and would be more consistent with Calypso multisite approach. A site that is not the last parameter should be documented edge case.

@ebinnion @enejb do you think we could reverse the url structure for people here?

aduth commented:
Does `:user` reference a WordPress.com user, or can it be the user on a Jetpack site? If it were the latter case, I don't think we can treat username as a global identifier.

rickybanister commented:
That's a great question @aduth. We are going to be thinking more about SSO and how we can improve our Jetpack people integration soon. In the meantime I think we can safely treat `:user` as a 'connected Jetpack user' meaning they have a .com account. Unconnected users will not work in a global scope at the moment, but I'd love to add an 'invite to connect' button next to their name in Calypso.

ebinnion commented:
User references the username on that site. So, this is a global identifier for all WP.com sites (since WP.com is a multisite installation), but it is a unique identifier for all single Jetpack installations.

> @ebinnion @enejb do you think we could reverse the url structure for people here?

Yes, we could. But, this would put us back in the same position where we can't support usernames with a `.` in them. And since the user ID is specific to each site for Jetpack sites, we could just as easily leave the people URL structure as it is.

Ultimately, it was very well be that people is the only edge case where this could happen, and perhaps we agree that we work around it by using IDs to identify users instead of usernames. But, with allowing the site fragment to be in more than one position, we may very well run into similar issues in the future in other sections.

enejb commented:
> I'd expect /people/edit/:user to work and show me all the sites a user belongs to. 
@mtias That only work if you had only .com sites. It wouldn't work for Jetpack sites. We currently support editing users that are not connected. 

I am not sure when an admin would be making global changes to a user and what would that mean. 
We could also use a key word like `global` to end up in that state. 
So something like `people/edit/global/user_id`. or `people/edit/network_id/user_id`

I am still not sure what user case that would be. Checking what roles this user has across multiple sites on a network?

mtias commented:
> I am still not sure what user case that would be. Checking what roles this user has across multiple sites on a network?

Calypso is multi-site by default, so the same way I can see my posts across different sites, seeing the user across multiple sites could have its uses. At the most basic it would be seeing which sites, among the ones you belong to, they also have access. More interestingly, once we have better collaboration tools there could be more interesting things here, like latest posts from the user across a network, comments, etc.
